### ISSUE TITLE
Backspace should not change the type of the trailing block (probably shouldn't merge it at all)

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#6680]
## 📝 Provide detailed reproduction steps (if any)

When having the following selection:

```
<paragraph>[Foo</paragraph>
<paragraph>Bar</paragraph>
<heading1>]Bom</heading1>
```

I expect to end with this after pressing Backspace:

```
<paragraph>[]</paragrph>
<heading1>Bom</heading1>
```

## Actual result

<img src="https://user-images.githubusercontent.com/156149/80364268-e7d8e280-8885-11ea-89ca-f762c7771f08.gif" width="450">

The format of the trailing block is changed to the format of the former block. Technically speaking, the heading is being merged into the first paragraph.

## UX

This is especially irritating when you try to delete e.g. paragraphs between 2 headings:

<img src="https://user-images.githubusercontent.com/156149/80364469-4a31e300-8886-11ea-8774-545592abee74.png" width="450">

I get the "Proposal" heading turned into a paragraph. What I'd expect is having the two paragraphs above deleted, one left and the heading not touched at all:

<img src="https://user-images.githubusercontent.com/156149/80364670-97ae5000-8886-11ea-935e-11ec72a9d988.png" width="450">

## Background

The trailing block problem is a well-known DOM selection representation issue. For instance, when you triple-click a line of text to select the entire line, the selection looks like this:

```
<paragraph>Foo</parargaph>
<paragraph>[Trippleclicked line</parargaph>
<paragraph>]Foo</parargaph>
```

Unfortunately:

*   We didn't plan to select the "Foo" block. It got selected by the browser because that's the only way how it represents a "new line is selected too" situation which is a different situation than selecting just the content of this block.
*   You don't see in Chrome (FF and Safari handle this better) that the end of line is selected too.

We already handled this situation in the `selection.getSelectedBlocks()` method, so e.g. when yoy apply a block format to this selection, it does not affect the "Foo" block.

It'd now be good to also handle this in `deleteContent()` by simply ignoring the trailing block if the selection starts at its beginning.

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by oleq** [comment]
What does it work like in other editors?

**Comment by wwalc** [comment]
Yeah, it happens to me too whenever I use shift + down arrow key to select blocks. Funny I never reported it...   
  
Just a note to be careful when working on it, so that the way how CKEditor 5 handles backspace did not break lists handling (if it may be in some way influenced by the fix implemented to solve this issue at all).

**Comment by niegowski** [comment]
## Selection cases

### 1\. Vertical selection (mouse or keyboard)

Mouse dragged vertically along the block elements (or keyboard <kbd>up</kbd> / <kbd>down</kbd> arrows) makes the selection expand to the start of next block element.   
 

![](https://user-images.githubusercontent.com/1232187/82923197-ea108880-9f7a-11ea-86fd-cfddde57869f.gif)

### 2\. Mouse tripple-click selection

Behaves exactly as vertical selection (includes start of next block element).

![](https://user-images.githubusercontent.com/1232187/82923612-85a1f900-9f7b-11ea-9911-734138f182fb.gif)

### 3\. Horizontal selection (mouse or keyboard)

Mouse dragged horizontally over the block elements (or keyboard <kbd>Shift</kbd> + <kbd>left</kbd> / <kbd>right</kbd> / <kbd>home</kbd> / <kbd>end</kbd> arrows) does not expand the selection to any other block element.   
 

![](https://user-images.githubusercontent.com/1232187/82923818-c437b380-9f7b-11ea-8481-ba551a57a2ea.gif)

### 4\. DeleteCommand

It's expanding the selection by one character in the direction of the command (<kbd>Backspace</kbd> - "backward", <kbd>Delete</kbd> - "forward") and then triggering `Model#deleteContent()` on that selection. 

![](https://user-images.githubusercontent.com/1232187/82924216-517b0800-9f7c-11ea-88b3-15a1b77b22a4.gif)

### 5\. EnterCommand

For the expanded selection is triggering `Model#deleteContent()` on it with `leaveUnmerged` option set to `true` if the selection is not from start to end of an element (horizontal selection of whole paragraph). `leaveUnmerged` keeps current block element in place but selection is moved to the start of next block element.

![](https://user-images.githubusercontent.com/1232187/82924832-34930480-9f7d-11ea-984c-0e93a0ea8d1b.gif)

### 6\. Paste from Clipboard

Content in clipboard does not contain trailing newline and after pasting it in front of some block they are merged.

![](https://user-images.githubusercontent.com/1232187/82926252-34940400-9f7f-11ea-8c0d-01ad65505d1b.gif)

**Comment by niegowski** [comment]
## Problems

### Cases 1 and 2 - the selection includes the start position of next block element

After deleting content the following block element should keep its type and attributes no matter whether it was <kbd>Delete</kbd> or <kbd>Backspace</kbd> (GDocs behave in this way).

![](https://user-images.githubusercontent.com/1232187/82928877-18926180-9f83-11ea-87b4-be55a35ffec8.gif)

### Case 3 - the selection doesn't include next block

This case works as in GDocs and seems reasonable.

![](https://user-images.githubusercontent.com/1232187/82928983-42e41f00-9f83-11ea-94ca-51fb6002f537.gif)

### Case 4 - DeleteCommand with empty block before the selection

Since `DeleteCommand` is expanding selection this case is very similar to cases 1 and 2 (here is no content, but in cases 1 and 2 after removing text from the selection this case is identical).

![](https://user-images.githubusercontent.com/1232187/82929003-4aa3c380-9f83-11ea-9fe8-a81c12b8fb95.gif)

### Case 5 - EnterCommand

Selection should not jump to the start position of following block element.

![](https://user-images.githubusercontent.com/1232187/82929294-b5ed9580-9f83-11ea-9701-0292f79cf435.gif)

### Case 6 - Paste from Clipboard

The clipboard data should contain trailing block if it was selected with it while copying/cutting.

![](https://user-images.githubusercontent.com/1232187/82929322-c0a82a80-9f83-11ea-8f55-42fd625dde89.gif)

**Comment by niegowski** [comment]
Described expected behavior is not sth GDocs do.

> When having the following selection:
> 
> ```
> <paragraph>[Foo</paragraph>
> <paragraph>Bar</paragraph>
> <heading1>]Bom</heading1>
> ```
> 
> I expect to end with this after pressing Backspace:
> 
> ```
> <paragraph>[]</paragrph>
> <heading1>Bom</heading1>
> ```

GDocs results with:

```
<heading1>[]Bom</heading1>
```

But title of this issue "**Backspace should not change the type of the trailing block**" is a valid point of the need of not merging those blocks as we do it currently and this is really annoying UX. 

How this currently works is a result of OT's preferred way of handling such situations, but as discussed in #4434 there is an option of "right-merging" those elements.

From cases described above there are 3 areas where we can improve:

*   `Model#deleteContent` should "merge right" in cases when left side of merging is empty and right one has content so that the properties of element with some content is preserved.
*   `EnterCommand` should not move caret to the next block if this wasn't intended/expected.
*   `Model#getSelectedContent` seems to be to aggressive when cleaning the document fragment while handling copy/cut.

**Comment by niegowski** [comment]
The `Model#deleteContent` issue is addressed in PR #7270.

I think there should be followups for:

*   `EnterCommand` 
*   `Model#getSelectedContent`

**Comment by niegowski** [comment]
After f2f talk with @Reinmar - there is one more use case:

Due to behavior of <kbd>Enter</kbd> (which is consistent with GDocs and Pages) after removing whole paragraphs hitting <kbd>Enter</kbd> creates another header block but in this scenario paragraph would be more reasonable so it seems it's better to not merge them. 

![](https://user-images.githubusercontent.com/1232187/83007833-863a9e00-a014-11ea-8e6b-35c8cc8efb81.gif)

So more user friendly behavior would be (despite it's not consistent with other editors):

![](https://user-images.githubusercontent.com/1232187/83008309-432cfa80-a015-11ea-91e8-2934b751fc86.gif)
